src/build_packages.sh

```diff
@@ -2,27 +2,58 @@
 set -e
 export XROAD=$(cd "$(dirname "$0")"; pwd)
 
-if [[ ! " $* " =~ " --packageonly " ]]; then
-  ./compile_code.sh "$@"
-fi
+HAS_DOCKER=""
+
+errorExit() {
+	echo "*** $*" 1>&2
+	exit 1
+}
+
+buildInDocker() {
+    test -n "$HAS_DOCKER" || errorExit "Error, docker is not installed/running."
+    echo "Building in docker..."
+    BUILD_IMAGE="`docker images | grep xroad-build | awk '{ print $1 }'`"
+    if [ "$BUILD_IMAGE" != "xroad-build" ]; then
+        echo "Build image 'xroad-build' not found, building it."
+        docker build -q -t xroad-build --build-arg uid=$(id -u) --build-arg gid=$(id -g) $XROAD/packages/docker-compile || errorExit "Error building build image."
+    fi
+    docker run --rm -v $XROAD/..:/workspace -w /workspace/src -u builder xroad-build  bash -c "./update_ruby_dependencies.sh && ./compile_code.sh -nodaemon" || errorExit "Error running build of binaries."
+}
+
+buildLocally() {
+    echo "Building locally..."
+    cd $XROAD || errorExit "Error 'cd $XROAD'."
+    ./compile_code.sh "$@" || errorExit "Error running build of binaries."
+}
 
 if command -v docker &>/dev/null; then
-    docker build -q -t xroad-deb-bionic "$XROAD/packages/docker/deb-bionic"
-    docker build -q -t xroad-deb-focal "$XROAD/packages/docker/deb-focal"
-    docker build -q -t xroad-rpm "$XROAD/packages/docker/rpm"
-    docker build -q -t xroad-rpm-el8 "$XROAD/packages/docker/rpm-el8"
+    HAS_DOCKER=true
+fi
+
+case "$1" in
+    --packageonly|-p) shift;;
+    --docker-build|-d) shift; buildInDocker "$@";;
+    *) shift; buildLocally "$@";;
+esac
```

Comment:
Maybe I would not use shift to remove the parameter. The scripts `build_packages.sh` and `compile_code.sh` seem to support parameter carrying so that you can e.g. call `build_packages.sh param1 param2` where param1 param2 are passed on as `compile_code.sh` parameters.

shift assumes that build-mode parameter (packageonly, docker-build, p, d, and any other parameter which means "build locally") is the first one. If order is different, a wrong parameter can get removed. Old script also meant calling `build_packages.sh param1` resulted in execution if `compile_code.sh param1`, but now it executes `compile_code.sh` which could be a breaking change for some. And also code compile_code.sh does not care about parameter order, maybe this script could use consistent approach.

I would probably make build-style parameter not required to be the first one, and just pass all parameters to compile_code without shifting. Either that, or document the behavior.

Reply:
You're right, the last shift should not be there. I have removed it. The first two are necessary, because the `--packageonly` or `--docker-build` arguments are solely for `build_packages.sh` and should not be passed to any other scripts.

Reply:
I made some suggestions that would make build-mode parameter order more flexible and make the handling implementation more consistent with how `compile_code.sh` does it:

- https://github.com/nordic-institute/X-Road/pull/726#pullrequestreview-555514097
- https://github.com/nordic-institute/X-Road/pull/726#discussion_r545856375